frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSquashAction.java

```diff
@@ -0,0 +1,180 @@
+      presentation.setEnabled(false);
+
+    } else if (branchName != null && numberOfCommits != null) {
+
+      if (numberOfCommits < 2) {
+        presentation.setDescription(
+            getNonHtmlString("action.GitMachete.BaseSquashAction.not-enough-commits")
+                .format(branchName, numberOfCommits.toString()));
+        presentation.setEnabled(false);
+      } else if (syncToParentStatus == InSyncButForkPointOff) {
+        presentation.setEnabled(false);
+        presentation.setDescription(getNonHtmlString("action.GitMachete.BaseSquashAction.fork-point-off")
+            .format(branchName));
+      } else {
+        val isSquashingCurrentBranch = getCurrentBranchNameIfManaged(anActionEvent)
+            .map(bn -> bn.equals(branchName)).getOrElse(false);
+        if (anActionEvent.getPlace().equals(ActionPlaces.ACTION_PLACE_CONTEXT_MENU) && isSquashingCurrentBranch) {
+          presentation.setText(getString("action.GitMachete.BaseSquashAction.text"));
+        }
+      }
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+
+    val project = getProject(anActionEvent);
+    val branchNameOption = getNameOfBranchUnderAction(anActionEvent);
+    val nonRootBranchOption = branchNameOption.flatMap(bn -> getManagedBranchByName(anActionEvent, bn))
+        .flatMap(b -> b.isNonRoot() ? Option.of(b.asNonRoot()) : Option.none());
+    val commitsOption = nonRootBranchOption
+        .map(b -> b.getCommits());
+    val parent = nonRootBranchOption
+        .flatMap(b -> b.getForkPoint()).getOrNull();
+    val syncToParentStatus = nonRootBranchOption.map(b -> b.getSyncToParentStatus()).getOrNull();
+    val gitRepository = getSelectedGitRepository(anActionEvent).getOrNull();
+    val branchName = branchNameOption.getOrNull();
+    val commits = commitsOption.getOrNull();
+
+    if (commits != null && gitRepository != null && parent != null && branchName != null
+        && syncToParentStatus != InSyncButForkPointOff) {
+      val isSquashingCurrentBranch = Option.of(gitRepository.getCurrentBranch()).map(b -> b.getName())
+          .flatMap(cbn -> branchNameOption.map(bn -> bn.equals(cbn))).getOrElse(false);
+      doSquash(project, gitRepository, parent, commits, branchName, isSquashingCurrentBranch);
+    }
+  }
+
+  @UIEffect
+  private void doSquash(Project project,
+      GitRepository gitRepository,
+      ICommitOfManagedBranch parent,
+      List<ICommitOfManagedBranch> commits,
+      String branchName,
+      boolean isSquashingCurrentBranch) {
+
+    @Data
+    // So that Interning Checker doesn't complain about enum comparison (by `equals` and not by `==`) in Lombok-generated `equals`
+    @SuppressWarnings("interning:not.interned")
+    class VcsCommitMetadataAndMessage {
+      private final List<VcsCommitMetadata> metadata;
+      private final String message;
+
+    }
+
+    val vcsCommitMetadataAndMessage = commits.foldLeft(
```

Comment:
Actually... I'd rather expect the **first** commit message to go into the resulting commit's **subject** (i.e. first line of message). It's usually more expected that the first commit message describes best what the entire branch/PR is about... subsequent commits bring in some fixes etc.

Reply:
Now AFAICS the commits are going **from the latest** to the earliest, i.e. latest commit subjects becomes the squashed commit subject

Reply:
yup. I made it so to stay consistent with IntelliJ. but thinking about it now I must agree that the order is strange. 

I will merge commit messages in the reverse order - earlier first